src/clients/FtpGateway.cc

```diff
@@ -402,6 +402,12 @@ Ftp::Gateway::loginParser(const SBuf &login, bool escaped)
     if (login.isEmpty())
         return;
 
+    if (!login[0]) {
+        debugs(9, 2, "WARNING: Ignoring FTP credentials that start with a NUL character");
```

Comment:
I am concerned that this debugging line was not present in the test output you have shared (quoted below). Are you sure the test is tickling the new code? Was the same test crashing Squid before?

> ```
> 2023/11/03 12:43:40.571 kid1| 9,2| src/clients/FtpClient.cc(810) writeCommand: ftp<< USER anonymous^M  
> 2023/11/03 12:43:40.642 kid1| 9,2| src/clients/FtpClient.cc(1127) parseControlReply: ftp>> 331 Please specify the password.   
> 2023/11/03 12:43:40.642 kid1| 9,2| src/clients/FtpClient.cc(810) writeCommand: ftp<< PASS anonymous@foo^M 
> 2023/11/03 12:43:43.571 kid1| 9,2| src/clients/FtpClient.cc(1127) parseControlReply: ftp>> 530 Login incorrect. 
> ...
> 2023/11/03 12:43:43.577 kid1| 11,2| src/http/Stream.cc(272) sendStartOfMessage: HTTP Client REPLY:                                                                                                                                       
> ---------                                                                                                                                                                                                                                                     
> HTTP/1.1 401 Unauthorized^M 
> ...
> X-Squid-Error: ERR_FTP_FORBIDDEN 530^M     
> ```

Reply:
Sorry for the late reply. It did tickle the code in question. It was admittedly rather stupid to not include that m(
```